tensorflow_recommenders_addons/dynamic_embedding/core/kernels/sparse_reshape_op.cu.cc

```diff
@@ -0,0 +1,173 @@
+#include "tensorflow/core/util/gpu_kernel_helper.h"
+
+namespace tensorflow {
+
+using GPUDevice = Eigen::GpuDevice;
+
+template <class IndexType>
+__global__ void SparseReshapeKernel(const IndexType* input_indices_in,
+                                    const int nnz, const int input_dim,
+                                    const IndexType* input_shape_in,
+                                    const IndexType* target_shape_in,
+                                    IndexType* output_indices,
+                                    const int output_dim) {
+#define RESHAPE_KERNEL_MAX_DIM 32
+  int64 input_strides[RESHAPE_KERNEL_MAX_DIM];
+  int64 output_strides[RESHAPE_KERNEL_MAX_DIM];
+#undef RESHAPE_KERNEL_MAX_DIM
+  // compute input strides
+  input_strides[input_dim - 1] = 1;
+  for (int i = input_dim - 2; i >= 0; i--) {
+    input_strides[i] = input_strides[i + 1] * input_shape_in[i + 1];
+  }
+  // compute output strides
+  output_strides[output_dim - 1] = 1;
+  for (int i = output_dim - 2; i >= 0; i--) {
+    output_strides[i] = output_strides[i + 1] * target_shape_in[i + 1];
+  }
+
+  GPU_1D_KERNEL_LOOP(idx, nnz) {
+    IndexType id = 0;
+#pragma unroll
+    for (int i = 0; i < input_dim; i++) {
+      id = id + input_strides[i] * input_indices_in[idx * input_dim + i];
+    }
+#pragma unroll
+    for (int i = 0; i < output_dim; i++) {
+      output_indices[idx * output_dim + i] = id / output_strides[i];
+      id = id % output_strides[i];
+    }
+  }
+}
+
+template <class IndexType>
+__global__ void DetermineOutputShapeKernel(const int input_dim,
+                                           const IndexType* input_shape,
+                                           const IndexType* target_shape,
+                                           const int output_dim,
+                                           IndexType* output_shape) {
+  int64 dense_size = 1;
+  for (int d = 0; d < input_dim; ++d) {
+    dense_size *= input_shape[d];
+  }
+  int64 product = 1;
+  int unknown_index = -1;
+  for (int d = 0; d < output_dim; ++d) {
+    const int64 size = target_shape[d];
+    if (size == -1) {
+      unknown_index = d;
+      output_shape[d] = 1;
+    } else {
+      product *= size;
+      output_shape[d] = size;
+    }
+  }
+
+  if (unknown_index != -1) {
+    const int64 missing = dense_size / product;
+    output_shape[unknown_index] = missing;
+  }
+}
+
+namespace functor {
+
+template <class IndexType>
+struct SparseReshapeImpl {
+ public:
+  static Status Compute(OpKernelContext* context,
+                        const IndexType* input_indices_in, const int nnz,
+                        const int input_dim, const IndexType* input_shape_in,
+                        const IndexType* target_shape_in,
+                        IndexType* output_indices, int output_dim,
+                        IndexType* output_shape) {
+    auto d = context->eigen_gpu_device();
+    DetermineOutputShapeKernel<<<1, 1, 0, d.stream()>>>(
```

Comment:
Have you compared the performance  of computing output_shape in the previous kernel or in every `SparseReshapeKernel` to avoid the cost of launching kernel. Since `DetermineOutputShapeKernel ` and `SparseReshapeKernel ` use the same stream, it will be sequential, whitch leading to same cost for every local thread. So I think it might be possible to avoid launch two kernels.

Reply:
When using `d.stream()` to compute the output_shape, the next kernel `SparseReshapeKernel` will wait until `DeterminedOutputSHapeKernel` finished. And thus the time cost might be same to compute in one kernel. And launch a kernel may take extra overhead.